torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1362,6 +1362,265 @@ def rbinary_magic_impl(self, other):
+            return x if x.is_int() else z3.ToInt(x)
+
+        # Implements Python division semantics.
+        # This is needed because Z3 integer division does not have the
+        # same semantics as Python true division.
+        def _div(self, num, divisor) -> z3.ArithRef:
+            return self._to_real(num) / self._to_real(divisor)
+
+        def _int(self, expr: int) -> z3.ArithRef:
+            return z3.IntVal(expr)
+
+        def _Symbol(self, expr: sympy.Symbol) -> z3.ArithRef:
+            return self._symbols[expr]
+
+        def _Rel(self, expr: sympy.Rel) -> z3.ArithRef:
+            opmap = {
+                "==": operator.eq,
+                "!=": operator.ne,
+                ">=": operator.ge,
+                ">": operator.gt,
+                "<=": operator.le,
+                "<": operator.lt,
+            }
+            assert isinstance(expr, (sympy.Eq, sympy.Ne, sympy.Ge, sympy.Gt, sympy.Le, sympy.Lt))
+            return opmap[expr.rel_op](self(expr.lhs), self(expr.rhs))
+
+        def _Add(self, expr: sympy.Add) -> z3.ArithRef:
+            return functools.reduce(operator.add, self(expr.args))
+
+        def _Mod(self, expr: sympy.Mod) -> z3.ArithRef:
+            p, q = [self._to_int(self(a)) for a in expr.args]
+            return p % q
+
+        # Turns the multiplication into a division, if there is any
+        # appropriate term.
+        #
+        # SymPy expresses division as a term in the multiplication that consists
+        # of a power expression where the exponent is negative.
+        #
+        # Therefore, we separate the whole multiplication into numerator and
+        # denominator. Then, we create a division node if necessary (i.e. there
+        # is any denominator not equal to 1).
+        #
+        # This separation was based on:
+        # sympy.printing.codeprinter.CodePrinter._print_Mul method.
+        def _Mul(self, expr: sympy.Mul) -> z3.ArithRef:
+            numerator_list = []
+            denominator_list = []
+
+            for item in expr.args:
+                if (
+                        item.is_commutative
+                        and isinstance(item, sympy.Pow)
+                        and item.exp.is_Rational
+                        and item.exp.is_negative
+                ):
+                    denominator_list.append(sympy.Pow(item.base, -item.exp))
+                else:
+                    numerator_list.append(item)
+
+            numerator = self._int(1)
+            if len(numerator_list) > 0:
+                numerator = functools.reduce(operator.mul, self(numerator_list))
+
+            if len(denominator_list) > 0:
+                denominator = functools.reduce(operator.mul, self(denominator_list))
+                return self._div(numerator, denominator)
+
+            return numerator
+
+        def _Pow(self, expr: sympy.Pow) -> z3.ArithRef:
+            base, exp = self(expr.args)
+            return base ** exp
```

Comment:
I didn't really find any specification of `pow` semantics for int or real theory online. I did a few experiments to try to understand it more. I designed the experiments so that they would always be `unsat`, unless the thing being tested (e.g. division-by-zero) is unspecified.

```python
def check(*args):
    s = Solver()
    for a in args: s.add(a)
    if s.check() == sat:
        return s.model()
    return s.check()

>>> a = Symbol("a", integer=True)
>>> b = Symbol("b", real=True)

# Division-by-zero: should be unsat
>>> check(a >= 0, b > 0, b / a < -5)
[a = 0, b = -1]

# Power with negative exponent + base zero: should be unsat
>>> check(a >= 0, a ** -8 < -5)
unsat

# Power with base zero + negative exponent: should be unsat
>>> check(a < 0, 0 ** a < -5)
unknown 

# Power with base negative + non-integer exponent: should be unsat
>>> check(a < 0, a ** 0.5 == 0)
unknown
```

- Only actual division-by-zero was `sat`, i.e. Z3 considered division-by-zero as an unspecified operation (function that can yield any number).
- Z3 does not consider `0 ** (-n)` as an unspecified operation.
- Going to the complex domain (e.g. `(-n) ** 0.5`) also doesn't result in `sat`.

So, maybe we don't need to deal with them? Let me know what you think.

Reply:
Then we should just deal with division, yep. Do leave either a link to this comment or a comment in the code, please.